airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlSourceOperations.kt

```diff
@@ -49,17 +58,65 @@ import io.airbyte.cdk.read.SelectNode
 import io.airbyte.cdk.read.SelectQuery
 import io.airbyte.cdk.read.SelectQueryGenerator
 import io.airbyte.cdk.read.SelectQuerySpec
+import io.airbyte.cdk.read.Stream
 import io.airbyte.cdk.read.Where
 import io.airbyte.cdk.read.WhereClauseLeafNode
 import io.airbyte.cdk.read.WhereClauseNode
 import io.airbyte.cdk.read.WhereNode
+import io.airbyte.cdk.read.cdc.DebeziumState
 import io.airbyte.cdk.util.Jsons
+import io.airbyte.integrations.source.mysql.cdc.MySqlDebeziumOperations
+import io.airbyte.integrations.source.mysql.cdc.MySqlPosition
 import io.micronaut.context.annotation.Primary
 import jakarta.inject.Singleton
+import java.time.OffsetDateTime
 
 @Singleton
 @Primary
-class MysqlSourceOperations : JdbcMetadataQuerier.FieldTypeMapper, SelectQueryGenerator {
+class MysqlSourceOperations :
+    JdbcMetadataQuerier.FieldTypeMapper, SelectQueryGenerator, JdbcAirbyteStreamFactory {
+
+    override val globalCursor: MetaField = MysqlCdcMetaFields.CDC_CURSOR
+
+    override val globalMetaFields: Set<MetaField> =
+        setOf(
+            MysqlCdcMetaFields.CDC_CURSOR,
+            CommonMetaField.CDC_UPDATED_AT,
+            CommonMetaField.CDC_DELETED_AT,
+            MysqlCdcMetaFields.CDC_LOG_FILE,
+            MysqlCdcMetaFields.CDC_LOG_POS
+        )
+
+    override fun decorateRecordData(
+        timestamp: OffsetDateTime,
+        globalStateValue: OpaqueStateValue?,
+        stream: Stream,
+        recordData: ObjectNode
+    ) {
+        recordData.set<JsonNode>(
+            CommonMetaField.CDC_UPDATED_AT.id,
+            CdcOffsetDateTimeMetaFieldType.jsonEncoder.encode(timestamp),
+        )
+        recordData.set<JsonNode>(
+            MysqlCdcMetaFields.CDC_LOG_POS.id,
+            CdcIntegerMetaFieldType.jsonEncoder.encode(0),
+        )
+        if (globalStateValue == null) {
+            return
+        }
+        val debeziumState: DebeziumState =
+            MySqlDebeziumOperations.deserializeDebeziumState(globalStateValue)
+        val position: MySqlPosition = MySqlDebeziumOperations.position(debeziumState.offset)
+        recordData.set<JsonNode>(
+            MysqlCdcMetaFields.CDC_LOG_FILE.id,
+            CdcStringMetaFieldType.jsonEncoder.encode(position.fileName),
+        )
```

Comment:
Maybe it's not important because values in these fields may not be used right now, but I wonder if this is correct. If I understand correctly, globalStateValue will only update in a batch, but this method get called for each record message. Thus it seems records in the same batch are going to have same debezium state and position.

Reply:
This method actually gets called only once per `PartitionReader`. Yes, records in the same batch are going to have same debezium state and position. To me, it makes sense that that would be the case, after all these records are not produced by debezium, so by doing so we're kind of pretending that they were produced along with the very last debezium record.

Reply:
I'm not certain, and I hope I'm not wrong, but I don't think the cursor value needs to be unique. It routinely isn't when doing user-defined incremental syncs. I asked about this a couple of weeks ago and at the time nobody had any definitive answer one way or the other.

Reply:
yeah that matches with my understanding too. thanks for clarifying